pkg/ensign/events.go

```diff
@@ -17,9 +23,63 @@ func (s *Server) Publish(stream api.Ensign_PublishServer) (err error) {
 	o11y.OnlinePublishers.Inc()
 	defer o11y.OnlinePublishers.Dec()
 
+	// Parse the context for authentication information
+	ctx := stream.Context()
+	claims, ok := contexts.ClaimsFrom(ctx)
+	if !ok {
+		// NOTE: this should never happen because the interceptor will catch it, but
+		// this check prevents nil panics and guards against regressions.
+		sentry.Fatal(ctx).Msg("no claims available in publish stream")
+		return status.Error(codes.Unauthenticated, "missing credentials")
+	}
+
+	// Verify that the user has permissions to publish.
+	if !claims.HasPermission(permissions.Publisher) {
+		log.Warn().Msg("attempt to open publisher stream without publisher permission")
+		return status.Error(codes.Unauthenticated, "not authorized to perform this action")
+	}
+
+	var projectID ulid.ULID
+	if projectID, err = ulids.Parse(claims.ProjectID); err != nil || ulids.IsZero(projectID) {
+		sentry.Warn(ctx).Err(err).Msg("could not parse projectID from claims")
+		return status.Error(codes.Unauthenticated, "not authorized to perform this action")
+	}
+
+	// Get the topic IDs this user is allowed to publish to
+	// TODO: make this more efficient and globalize to prevent inconsistencies
```

Comment:
Is there a followup story for this?

Reply:
No, but my next push will take care of it.